Assignment3/Q1.py

```diff
@@ -1,128 +1,125 @@
+                     visited by the search the corresponding 'visited entry' is set to True and otherwise it is set to
+                     False
+            start: a pair of integers, the entry ('vertex') we start our search from
+            word_list: a list of strings, list of words that can be formulated from a character grid and are
+                       in a dictionary
             prefix: a string, contains a word 'accumulated' so far, in a recursive call prefix is modify in order
                     to perform a DFS search, we terminate early if isPrefix(string) on a current prefix returns False
-            words: a list of strings, list of words that can be formulated from a character grid and are
-                   in a dictionary
-            path: list of integer tuples (similar to path in DFS search) we keep here an information about the vertices
-                  visited so far in a search, it prevents visiting the same cell in a grid twice in a single word search
 
         Returns:
-            a list of strings, list of words that can be form using characters in a grid, stating from start vertex
-            (corresponding to some character in a grid) that are in a dictionary.
+            it is a void function, no value is returned but the function modifies its word_list argument which
+            accumulates the words found in a grid.
     """
-    # Get coordinates of the starting vertex in a search.
+    # Get coordinates of the starting entry in a search.
     (i, j) = start
 
     # Add a corresponding character to the prefix 'accumulated' so far.
-    # NOTE: We are case insensitive in this solution
-    prefix = prefix + grid[i][j].upper()
+    prefix = prefix + grid[i][j]
 
-    # Add a vertex to a path to prevent using the same grid cell twice in a search for single word.
-    path = path + [start]
+    # Indicate that the entry has been visited by search.
+    visited[i][j] = True
 
     # Early terminate - cut off the search tree branch if the prefix so far is not a prefix in a dictionary.
-    if not dictionary.isPrefix(prefix):
-        return words
-
-    # Add a word to word list if it is a word in a dictionary.
-    if dictionary.isWord(prefix):
-        words = words + [prefix]
-
-    # Recursive call, DFS-like search but searches ALL paths from a vertex
-    for edge in graph[start]:
-        if edge not in path:
-            words = word_search_recursive(dictionary, grid, graph, edge, prefix, words, path)
-    return words
-
+    if dictionary.isPrefix(prefix):
+        # Add a word to word list if it is a word in a dictionary.
+        if dictionary.isWord(prefix):
+            # Do not allow duplicates in a list of words.
+            if prefix not in word_list:
+                word_list += [prefix]
+
+        # Get rows and columns of a grid.
+        rows = len(grid)
+        columns = len(grid[0])
+
+        # Two arrays to keep track of adjacent entries.
```

Comment:
These are essentially constants. At the moment they are created each time we hit this and therefore O(N^4) times.

You can instead define top-level constants, e.g.:
```python
# Two arrays to keep track of adjacent entries.
_ROW_ADJACENT = [-1, -1, -1, 0, 0, 1, 1, 1]
_COLUMNS_ADJACENT = [-1, 0, 1, -1, 1, -1, 0, 1]
```
The name started with `_` because these are meant to be private to this module.
You can add your comment on top of them.

As a plus, you can define a single constant containing pairs of row and column deltas instead: this means that each pair is clearly identified (right now, you need to understand both variables together).